src/dmd/escape.d

```diff
@@ -420,15 +420,20 @@ bool checkParamArgumentEscape(Scope* sc, FuncDeclaration fdc, Parameter par, Exp
         }
     }
 
+    if (!sc.func)
+        return result;
+
     foreach (Expression ee; er.byexp)
     {
-        if (sc.func && sc.func.setUnsafe())
+        if (!par)
         {
-            if (!gag)
-                error(ee.loc, "reference to stack allocated value returned by `%s` assigned to non-scope parameter `%s`",
-                    ee.toChars(),
-                    par ? par.toChars() : "this");
-            result = true;
+            result = result || sc.func.setUnsafe(gag, ee.loc,
+                "reference to stack allocated value returned by `%s` assigned to non-scope parameter `this`", ee);
+        }
+        else
+        {
+            result = result || sc.func.setUnsafe(gag, ee.loc,
+                "reference to stack allocated value returned by `%s` assigned to non-scope parameter `%s`", ee, par);
```

Comment:
Hmm, setUnsafe's last 2 parameters should be `const(char)*`. There is no point in actually passing the object since `toChars` is going to be called anyway. That way we can avoid such code duplication.

Reply:
Also I don't think duplication of error messages is that problematic in general. The plus side is it makes them easier to grep for than when they're full of dynamically generated `%s` parts.